src/plots/cartesian/category_order_defaults.js

```diff
@@ -8,32 +8,25 @@
 
 'use strict';
 
-var layoutAttributes = require('./layout_attributes');
 
 module.exports = function handleCategoryOrderDefaults(containerIn, containerOut, coerce) {
+    if(containerOut.type !== 'category') return;
 
-    if(containerIn.type !== 'category') return;
+    var arrayIn = containerIn.categoryarray,
+        orderDefault;
 
-    var validCategories = layoutAttributes.categoryorder.values;
+    var isValidArray = (Array.isArray(arrayIn) && arrayIn.length);
```

Comment:
I also like the brevity of `a.length` rather than `a.length > 0` but assume that most places want the more explicit version :-)